### PR TITLE
272 - Create wizard complete with multipart upload functionality

### PR DESCRIPTION
**Pull request by deeleman**
This PR finalizes the implementation of the create api connector wizard, plus introduces the lazy loading pattern we should follow from now on to reduce http calls to the backend by leveraging the NGRX store when rendering data on components.

Although not an intrinsic part of the create wizard, the list and detail pages have also been refactored to leverage the Store functionality when rendering data.

The detail page is not complete since we will require #898 to be delivered first, in order to have a consistent data graph for the summary component in the detail view.

I noticed the list page still does not implement delete functionality.

### CODE REVIEW

File: app/ui/src/app/customizations/api-connector/api-connector.models.ts [hunk 1]
```diff
@@ -1,98 +1,64 @@
...
-  specification: {
-    kind: string;
-    displayName: string;
-    required: boolean;
-    type: string;
-    javaType: string;
-    description: string
+export interface ApiConnectorData {
+  id?: string;
+  actionsSummary?: {
```
Comment (kahboom) [inline app/ui/src/app/customizations/api-connector/api-connector.models.ts]: Wow this changed quite a bit.

File: app/ui/src/app/core/providers/api-http-provider.service.ts [hunk 2]
```diff
@@ -45,59 +47,74 @@ export class ApiHttpProviderService extends ApiHttpService {
...
+      .catch(error => Observable.throw(this.catchError(error)));
   }
 
   delete<T>(endpoint: string | any[]): Observable<T> {
     const { endpointKey, endpointParams } = this.deconstructEndpointParams(endpoint);
     const url = this.getEndpointUrl(endpointKey, ...endpointParams);
-    return this.httpClient.delete<T>(url);
+    return this.httpClient
+      .delete<T>(url)
+      .catch(error => Observable.throw(this.catchError(error)));
```
Comment (kahboom) [inline app/ui/src/app/core/providers/api-http-provider.service.ts]: Interesting that we weren't catching those before..

File: app/ui/src/app/customizations/api-connector/api-connector.reducer.ts [hunk 3]
```diff
@@ -48,10 +61,69 @@ export function apiConnectorReducer(state = initialState, action: any): ApiConne
...
+        hasErrors: false,
+        errors: []
+      };
+    }
+
     case ApiConnectorActions.FETCH: {
       return {
         ...state,
         loading: true,
+        loaded: false,
```
Comment (kahboom) [inline app/ui/src/app/customizations/api-connector/api-connector.reducer.ts]: We are not returning a default state in case there are no changes?

Comment (kahboom) [reply app/ui/src/app/customizations/api-connector/api-connector.reducer.ts]: Ah nevermind, I see it here: https://github.com/deeleman/syndesis/blob/0c76af5ce20716b0253338c02615f194a77f34fd/app/ui/src/app/customizations/api-connector/api-connector.reducer.ts#L155 👍

File: app/ui/src/app/customizations/api-connector/api-connector.service.ts [hunk 4]
```diff
@@ -1,23 +1,43 @@
...
       .setEndpointUrl('getApiConnectorDetails', { id })
       .get<ApiConnectorData>();
   }
 
-  list(): Observable<ApiConnectors> {
+  getApiConnectorList(): Observable<ApiConnectors> {
     return this.apiHttpService
       .setEndpointUrl('getApiConnectorList', { template: 'swagger-connector-template'})
-      .get<{ items: ApiConnectors }>() // TODO: Provide a better typing T here
+      .get<{ items: ApiConnectors }>() // TODO: Provide a better <T> typing here
```
Comment (kahboom) [inline app/ui/src/app/customizations/api-connector/api-connector.service.ts]: TODO: No more TODOS! 😭

File: app/ui/src/app/customizations/customizations.module.ts [hunk 5]
```diff
@@ -25,7 +25,8 @@ const routes: Routes = [
       },
       {
         path: 'api-connector',
-        component: ApiConnectorListComponent
+        component: ApiConnectorListComponent,
+        canActivate: [ApiConnectorLazyLoaderGuard]
```
Comment (kahboom) [inline app/ui/src/app/customizations/customizations.module.ts]: Wow we even have our own guard now!

File: app/ui/src/app/customizations/tech-extensions/import/tech-extension-import.component.ts [hunk 6]
```diff
@@ -9,10 +10,10 @@ import {
   ParsedResponseHeaders
 } from 'ng2-file-upload';
 import { NotificationType } from 'patternfly-ng';
-import { Extension } from '../../../model';
-import { ExtensionStore } from '../../../store/extension/extension.store';
-import { NotificationService } from 'app/common/ui-patternfly/notification-service';
-import {Observable} from "rxjs/Observable";
+
+import { Extension } from '@syndesis/ui/model';
```
Comment (kahboom) [inline app/ui/src/app/customizations/tech-extensions/import/tech-extension-import.component.ts]: 👍

**Comment by kahboom** [comment]
LGTM, surely we need to resolve the Delete for List Page, but that's a separate PR altogether. I really like where this is headed.

File: app/ui/src/app/customizations/customizations.module.ts (hunk 5 above)
Comment (deeleman) [reply app/ui/src/app/customizations/customizations.module.ts]: Yep, that guard takes care of checking if the ApiConenctors have been fetched from the API already and will trigger the fetch action if not, which will set the overall state in "loading", so when the user makes it to ANY page requiring that data, the loading state will be already on and the connectors will be rendered once loaded. Next time you return to that same page, no fetch (hence no http call) action will be triggered.

Try it out yourself: Go to the API Connectors page with the Networks tab open. You'll see a http request being made. Go back, then return to the page: No more http requests are triggered, since the information is now fetched from the store.

Try this out now: Go straight to the any connector detail view. Hard reload the page and see how you're fetching ALL the connectors, including the one you need. Now click on the lsit link and see that you're no longer requesting anything form the backend.

Last test: Go to the wizard and create a new connector (use this URL for fetching a awagger file: `http://petstore.swagger.io/v2/swagger.json`). Complete the steps and see how the connectors are re-fetched once the create process is done. Next time you access the list page, you're newly created connector will be already there.

Bottom line is that, by doing this, we:

* Reduce the backend overhead by minimizing HTTP calls
* Pages speed becomes blazing fast since all the information is reactively returned from the store.

Probs some people will thing the above is trivial, but think for a minute this site had millions of uniques a day hammering the API. This kinda architecture ensures site resilience and stability.

File: app/ui/src/app/customizations/api-connector/api-connector.service.ts (hunk 4 above)
Comment (deeleman) [reply app/ui/src/app/customizations/api-connector/api-connector.service.ts]: TODO: No more technical debt. I agree that this is not ideal, but tracking stuff that needs to be catered (and cannot be fixed at this stage) is a good way to embrace the [The boy Scout Rule](http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule) and prevent the introduction of more technical debt in the future.

Also, doing this is a good way to prevent other programmers like: "Hey, this needs to be fixed. Can you do it now?"

File: app/ui/src/app/customizations/api-connector/api-connector.reducer.ts (hunk 3 above)
Comment (deeleman) [reply app/ui/src/app/customizations/api-connector/api-connector.reducer.ts]: > We are not returning a default state in case there are no changes?

Certainly we do! But changes is not the right term here, Rachel. Our reducer handles ACTIONS. Therefore, we do return a default state (which is the current state snapshot) when the action that executes this reducer is not intercepted by any action handler in the switch/case block.

File: app/ui/src/app/core/providers/api-http-provider.service.ts (hunk 2 above)
Comment (deeleman) [reply app/ui/src/app/core/providers/api-http-provider.service.ts]: Actually the application is not very resilient to backend exceptions, so it breaks half of the times when something goes wrong upon requesting data.

By combining catch operators propagated from the ApiHttpService with catch operators in RxJS streams (usually located at the related Effects middleware class) which funnel the caught error to a a particular XXXX_FAIL action, we ensure the application is resilient and never gets in a bad state that prevents the user from keep browsing.

This also allows to leverage generic error components and display proper error messages in the components with minimum boilerplate, since all instances of state feature a common `hasErrors::Boolean` and `errors::Array<Error>` properties.

File: app/ui/src/app/customizations/api-connector/api-connector.models.ts (hunk 1 above)
Comment (deeleman) [reply app/ui/src/app/customizations/api-connector/api-connector.models.ts]: These models are far from being convenient and Frontend/Angular-friendly, but this is the best I could get to map the information we need from the backend in a simple and lean fashion.

Backend and Frontend will have to compromise in a better way. Otherwise this app will not scale.

File: app/ui/src/app/customizations/tech-extensions/import/tech-extension-import.component.ts (hunk 6 above)
Comment (deeleman) [reply app/ui/src/app/customizations/tech-extensions/import/tech-extension-import.component.ts]: This was unrelated to this issue, but as I pointed out earlier, any software engineer that worths this title should observe the Boy Scout Rule with no exception, and I found this when linting the PR.

File: app/ui/src/app/customizations/customizations.module.ts (hunk 5 above)
Comment (deeleman) [reply app/ui/src/app/customizations/customizations.module.ts]: Also, it's worth mentioned the change applied in the overall component controller implementation. Relying on a Store reduces the boilerplate to the max, since components become mere consumers of the store and action dispatchers, not requiring any further logic in their implementation.

File: app/ui/src/app/customizations/api-connector/api-connector.reducer.ts (hunk 3 above)
Comment (kahboom) [reply app/ui/src/app/customizations/api-connector/api-connector.reducer.ts]: Correct, I'm terrible with terminology.

File: app/ui/src/app/customizations/tech-extensions/import/tech-extension-import.component.ts (hunk 6 above)
Comment (kahboom) [reply app/ui/src/app/customizations/tech-extensions/import/tech-extension-import.component.ts]: I don't believe in separate PRs for cleanups like this, or they'll never get done. 👍
